File: workflows/python/workflowscript.py
# First import the packages that will be needed  
import boto3
import pickle
import os
import tempfile
import snowflake.connector

import pandas as pd

from snowflake.connector.pandas_tools import write_pandas
from sklearn.linear_model import LinearRegression

# Set some variables up to use them later
bucket = os.environ['DATA_LAKE'] # For the data lake connection
filepath = os.environ['DATA_LAKE_ROOT_PATH'] + 'datascience/' # For the datalake connection
filename = 'predictions.csv' # Filename for the prediction data 
key_model_save = os.environ['DATA_LAKE_ROOT_PATH'] + "datascience/model.pkl" # Model name


## This function connects you to Snowflake
def connect_to_db():
    conn = snowflake.connector.connect(
            user=os.environ['SNOWFLAKE_USERNAME'],
            password=os.environ['SNOWFLAKE_PASSWORD'],
            account=os.environ['SNOWFLAKE_ACCOUNT'],
            warehouse=os.environ['SNOWFLAKE_WAREHOUSE'],
            database=os.environ['SNOWFLAKE_DATABASE'], 
            schema=os.environ['SNOWFLAKE_SCHEMA'])
    return conn


def read_from_db(connection_str, query):
    cur = conn.cursor()
    cur.execute(query)
    df = cur.fetch_pandas_all()
    return df

# Printing statements so we have some logs
print("Querying Snowflake data")

# Query to get all of our raw data from STAGE
query = """
select * from STAGE.HOUSEPRICES
"""
conn = connect_to_db()
df = read_from_db(conn, query)

###
Pre process our data. Here we are:
- removing the prices that are below zero
- removing categorical variables
You can change this script if you want to remove some more values
###

print("Pre-processing data")
df = df[df['price'] > 0]

data = df.drop(['date',
                'street', 
                'city', 
                'statezip',
                'country', 
                'PEAKAUDITCREATEDAT', 
                'REQUEST_ID'], axis=1)

# Building and fitting a Linear Regression model. You can change this
print("Build Model and add predictions to dataset")
model = LinearRegression().fit(data.drop('price', axis=1), data['price'])

# Prediciting on the data we have and appending the column to the orginal df 
predictions = model.predict(data.drop('price', axis=1))
df['predictions'] = predictions

# Altering the datatypes of df, ready to be saved into Snowflake
print("Changing datatypes as table in Snowflake has integers")
df['price'] = df['price'].astype(int)
df['predictions'] = df['predictions'].astype(int)
df['bedrooms'] =  df['bedrooms'].astype(int)
df['bathrooms'] =  df['bathrooms'].astype(int)
df['floors'] =  df['floors'].astype(int)

# Removing some fields we dont need 
df = df.drop(['PEAKAUDITCREATEDAT', 'REQUEST_ID'], axis=1)

# Changing the columns to upper case for Snowflake
df.columns = map(lambda x: str(x).upper(), df.columns)

# Creating some querys to create, delete and copy the prediction from our datalake
print("Querys to create, delete and copy predictions from S3")
create_table_query = """
CREATE OR REPLACE TABLE PUBLISH.HOUSEPRICE_PREDICTIONS (
date varchar(256),
  price integer,
  bedrooms integer,
  bathrooms integer,
  sqftliving integer,
  sqftlot integer,
  floors integer,
  waterfront integer,
  view integer,
  condition integer,
  sqftabove integer,
  sqftbasement integer,
  yrbuilt integer,
  yrrenovated integer,
  street varchar(256),
  city varchar(256),
  statezip varchar(256),
  country varchar(256),
  predictions integer    
)
"""

delete_query = "delete from PUBLISH.HOUSEPRICE_PREDICTIONS"

# Using our connection string to execture these queries
conn.cursor().execute(create_table_query)

conn.cursor().execute(delete_query)

print("Writing predictions data to Snowflake")

success, _, nrows, output = write_pandas(
            conn,
            df,
            table_name="HOUSEPRICE_PREDICTIONS",
            schema="PUBLISH",
            quote_identifiers=False
        )

print("Number of successful rows {}".format(nrows))

# Saving our model to our datalake to use later
print("Save model to s3")
s3 = boto3.client("s3")
with tempfile.TemporaryFile() as fp:
    pickle.dump(model, fp)
    fp.seek(0)
    s3.upload_fileobj(fp, 
                      Bucket=os.environ['DATA_LAKE'],
                      Key=key_model_save)
                      

File: workflows/r/workflowscript.R
# Import packages that you need 
library(dplyr)
library(odbc)
library(DBI)
library(tidymodels)

# Setting some variables
bucket = Sys.getenv('DATA_LAKE') # For the data lake connection
filepath = paste0(Sys.getenv('DATA_LAKE_ROOT_PATH') , 'datascience/') # For the data lake connection
modelname = 'model.rds' # Model name

# Connection string for Snowflake
print("Connecting to Snowflake")
conn = dbConnect(odbc::odbc(),
                 driver = '/usr/lib/snowflake/odbc/lib/libSnowflake.so',
                 server = paste0('peak-', Sys.getenv('TENANT'),'.snowflakecomputing.com'),
                 database = Sys.getenv('SNOWFLAKE_DATABASE'),
                 uid = Sys.getenv('SNOWFLAKE_USERNAME'),
                 pwd = Sys.getenv('SNOWFLAKE_PASSWORD')
)

# Query the raw data and filter out house prices 
print("Get data and filter price")
df = dbGetQuery(conn, "select * from STAGE.HOUSE_PRICES") %>%
  filter(price > 0)

# Removing categorical variables
print("Only select numerical data")
data = df %>%
  select(-date,
         -city,
         -PEAKAUDITCREATEDAT, 
         -REQUEST_ID)

# Building and fitting Linear Regression model. You can change this
print("Build model")
linreg_reg_spec <- 
  linear_reg() %>% 
  set_engine("lm")

linreg_reg_fit <- linreg_reg_spec %>% fit(price ~ ., data = data)

# Predicting on our data and creating a column called predictions
predictions = df %>%
  add_column(predict(linreg_reg_fit, data))  

names(predictions)[18] <- 'predictions'

# Save the model to the data lake using the AWS CLI functionality
print("Save model to S3")
saveRDS(linreg_reg_fit, 'model.rds')
system(glue::glue("aws s3 cp 'model.rds' 's3://{bucket}/{filepath}{modelname}'"))

# Changing the data types so that we can save the data to Snowflake
print("Altering the data")
predictions = predictions %>%
  mutate(price = as.integer(price), 
          predictions = as.integer(predictions), 
          bedrooms = as.integer(bedrooms), 
          bathrooms = as.integer(bathrooms), 
          floors = as.integer(floors) ) %>%
  select(-PEAKAUDITCREATEDAT, 
         -REQUEST_ID)

# Changing the column names to upper case so that we can save to Snowflake
names(predictions) <- toupper(names(predictions))

# Creating two queries to create and delete from the tables  
print("Save outputs to Snowflake")
create_table_query = "
CREATE OR REPLACE TABLE PUBLISH.HOUSEPRICE_PREDICTIONS (
date varchar(256),
  price integer,
  bedrooms integer,
  bathrooms integer,
  sqftliving integer,
  sqftlot integer,
  floors integer,
  waterfront integer,
  view integer,
  condition integer,
  sqftabove integer,
  sqftbasement integer,
  yrbuilt integer,
  yrrenovated integer,
  city varchar(256),
  predictions integer    
)

"
delete_query = "delete from PUBLISH.HOUSEPRICE_PREDICTIONS"

# Using DBI to run the queries on Snowflake
dbSendQuery(conn, create_table_query)
dbSendQuery(conn, delete_query)

print("Saving data to Snowflake")

# Appending to the Snowflake Table
dbAppendTable(conn, SQL("PUBLISH.HOUSEPRICE_PREDICTIONS"), predictions)
print("Complete")
